src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -298,7 +298,7 @@ export class HttpConnection implements IConnection {
     }
 
     private async getNegotiationResponse(url: string): Promise<INegotiateResponse> {
-        const headers = {};
+        const headers = this.httpClient.getHeaders();
```

Comment:
Because you're applying the headers from both the request and the httpclient in `HttpClient.send()` you don't need to call `getHeaders()` in any of these places. In fact, I think we can get rid of that method entirely.

Reply:
I think the one place we might still need `getHeaders()` is in WebSocketTransport.ts since that doesn't use `HttpClient.send()`.

Reply:
Nice catch!

Reply:
ServerSentEventsTransport.ts is the other exception, but like WebSocketTransport.ts it only matters on node, since you cannot set WebSocket or EventSource headers in the browser.

Reply:
> Because you're applying the headers from both the request and the httpclient in `HttpClient.send()` you don't need to call `getHeaders()` in any of these places. 

It is true only for `DefaultHttpClient`. But when I want to pass a different `httpClient` it works as expected - SignalR gets the headers by calling `getHeaders()` method. For example:
```ts
class MyHttpClient extends HttpClient {
  async send(request: HttpRequest) {
    // Do not add any headers here
  }
  getHeaders() {
    return { "X-API-KEY": "SECRET" };
  }
}
const connection = new signalR.HubConnectionBuilder()
    .withUrl("/chat", { httpClient: new MyHttpClient() })
    .build();
```

Reply:
> SignalR gets the headers by calling `getHeaders()` method

That's the wrong extension point, users should be passing headers via the new option. If they want to do special things in their custom http client then they can do that in send.

Reply:
I assume this is also why you are asking to add HttpClient.setHeaders() in https://github.com/dotnet/aspnetcore/pull/19227/files#r384010195 @BrennanConroy. That way headers can be configured once instead of every place the SignalR client makes an HTTP request. I might call it HttpClient.setDefaultHeaders() instead, but I think I agree with this approach.

I can also understand wanting to explicitly set all the headers for each request to make sure this change doesn't require *any* changes to existing HttpClients. I'm fine with this approach too. But in that case, there's no reason to pass the headers to the DefaultHttpClient constructor, since those same headers will be passed in to every call to HttpClient.send() via the HttpRequest parameter anyway.

Whichever approach we use, I don't see the reason behind HttpClient.getHeaders(). These headers are fully defined by IHttpConnectionOptions.headers. The MessageHeaders can be referenced directly from wherever HttpClient.send is being called. If it makes things easier, we could use a private HttpClientAndHeaders object to pass both around together, but I don't think getHeaders() should be added to the abstract base class.

Reply:
> I think the one place we might still need getHeaders() is in WebSocketTransport.ts since that doesn't use HttpClient.send().

@halter73 It was the original idea to attach headers to the abstract class `HttpClient` by adding the `getHeaders()` method. So no additional changes are required to existing HttpClients.